[MLIR] Fix VectorEmulateNarrowType constant op mask bug

lialan opened the pull request:
This commit adds support for handling mask constants generated by the `arith.constant` op in the `VectorEmulateNarrowType` pattern. Previously, this pattern would not match due to the lack of mask constant handling in `getCompressedMaskOp`.

The changes include:

1. Updating `getCompressedMaskOp` to recognize and handle `arith.constant` ops as mask value sources.

2. Handling cases where the mask is not aligned with the emulated load width. The compressed mask is adjusted to account for the offset.

Limitations:
- The arith.constant op can only have 1-dimensional constant values.

Resolves: #115742

lialan committed:
[MLIR] Fix VectorEmulateNarrowType constant op mask bug

This commit adds support for handling mask constants generated by the
`arith.constant` op in the `VectorEmulateNarrowType` pattern. Previously, this
pattern would not match due to the lack of mask constant handling in
`getCompressedMaskOp`.

The changes include:

1. Updating `getCompressedMaskOp` to recognize and handle `arith.constant` ops as
   mask value sources.

2. Handling cases where the mask is not aligned with the emulated load width.
   The compressed mask is adjusted to account for the offset.

Limitations:
- The arith.constant op can only have 1-dimensional constant values.

Resolves: #115742

Signed-off-by: Alan Li <me@alanli.org>

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
       newMask = rewriter.create<arith::ConstantOp>(loc, newMaskType, denseAttr);
     }
+  } else if (constantOp) {
+    assert(shape.size() == 1 && "expected 1-D mask");
```
Why do we have the assertion in this branch, but not others?

lialan replied:
@hanhanW Because it other branch it can handle multiple dimensional masks.

But should we simply only deal with 1d vectors in this pass?

hanhanW replied:
In this case, we can `return failure()` and add a TODO comment saying that it is not yet implemented. We should not crash in debug build and I believe the patterns can recover the state for this case.

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+    // region. For example, in the case of using byte-size for emulation, given
+    // the following mask:
+    //
+    //   %mask = vector.constant_mask [0, 1, 0, 1, 0, 0] : vector<6xi2>
```
I'm confused about `vector.constant_mask` here. The comment looks off to me because of the definition of the vector.constant_mask op: https://github.com/llvm/llvm-project/blob/00f2989f98520c401f0ab544a3dc766ed83785c0/mlir/include/mlir/Dialect/Vector/IR/VectorOps.td#L2347-L2383

Do you mean `arith.constant`? Also, should the type be i1? To simplify the comment, we can omit the op name and types, which looks cleaner to me. The idea of having the comment is to help people understand the logic. IMO, we don't need the concrete ops at this scope. So it could be something like `%mask = [1, 1, 0, 0, 0, 0]` (and it is how we document in the above function comment).

lialan replied:
You are right, and I am embarrassed by missing this part while doing self proof reading. Thanks!

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+    // be compressed.
+    SmallVector<bool> maskValues(intraDataOffset, false);
+    if (auto denseAttr =
+            mlir::dyn_cast<DenseIntElementsAttr>(constantOp.getValue())) {
```
nit: drop `mlir::`
style nit: prefer using early-exit to simplify the code, so it can be something like:

```cpp
auto denseAttr = dyn_cast...
if (!denseAttr)
  return failure();
// The rest code in the if branch..
``` 

https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

lialan replied:
ahh something I missed

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+    SmallVector<bool> maskValues(intraDataOffset, false);
+    if (auto denseAttr =
+            mlir::dyn_cast<DenseIntElementsAttr>(constantOp.getValue())) {
+      for (auto value : denseAttr.getValues<bool>()) {
+        maskValues.push_back(value);
+      }
```
I'm not sure if it works or not -- can we use the append method, like `maskValues.append(denseAttr.getValues<bool>())`?

lialan replied:
we could, just that it does not take `range` type, but iterators works!

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+      for (auto value : denseAttr.getValues<bool>()) {
+        maskValues.push_back(value);
+      }
+      while (maskValues.size() < numElements * scale) {
+        maskValues.push_back(false);
+      }
```
I think the size of maskValues should always be less than or equal to `numElements * scale`? If so, perhaps we can have an assertion and write `maskValues.resize(numElements * scale, false)`.

lialan replied:
Good suggestion!

hanhanW commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir:
```diff
@@ -624,3 +624,28 @@ func.func @vector_maskedstore_i4_constant_mask(
+}
+
+// CHECK: func @vector_maskedload_i4_arith_constant(
+// CHECK-SAME:   %[[PASSTHRU:[a-zA-Z0-9]+]]: vector<8xi4>) -> vector<8xi4> {
```
You don't need to capture the types (and the rest of function) because we have regular expression in the matching (i.e., [a-zA-Z0-9]+`). The trick allow us to omit the type when we don't care about it in the lit test.

```suggestion
// CHECK-SAME:   %[[PASSTHRU:[a-zA-Z0-9]+]]
```

hanhanW commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+      }
+    } else {
+      return failure();
+    }
+    // Compressing by combining every `scale` elements:
```
I'd suggest to have a blank line right before this comment. Because it is clearer that the code is not part of the above big chunk of comment. The above comment does not explain any compression. IMO, we should either expand the comment or just make a clear line between the codes. The latter one looks okay to me because this is what the method aims to do (and it's already documented in function comments).

hanhanW reviewed (commented):
Overall looks good to me, just some nits about style/comments/tests.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+    // region. For example, in the case of using byte-size for emulation, given
+    // the following mask:
+    //
+    //   %mask = vector.constant_mask [0, 1, 0, 1, 0, 0] : vector<6xi2>
```
This should be
```mlir
    //   %mask = vector.constant_mask [0, 1, 0, 1, 0, 0] : vector<6xi1>
```
right? (`i1` instead of `i2`).

lialan replied:
yeah, my mistake, and I updated them.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+    // back so that its length is multiple of `scale` (and the total coverage
+    // size is mulitiple of bytes):
+    //   %new_mask = vector.constant_mask [0, 0, 1, 0, 1, 0, 0, 0] :
+    //   vector<8xi2>
```
This should be `vector<8xi1` rather than `vector<8xi2>`, right? Also, don't we need two `i8` elements to load 6 x `i2`?

lialan replied:
same as above.

banach-space commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir:
```diff
@@ -624,3 +624,28 @@ func.func @vector_maskedstore_i4_constant_mask(
+// CHECK-SAME:   %[[PASSTHRU:[a-zA-Z0-9]+]]: vector<8xi4>) -> vector<8xi4> {
+// CHECK: %[[ALLOC:.+]] = memref.alloc() : memref<12xi8>
+// CHECK: %[[MASK:.+]] = arith.constant dense<[false, true, true, true, true, false, false, false]> : vector<8xi1>
+// CHECK: %[[MASK2:.+]] = arith.constant dense<[true, true, true, false]> : vector<4xi1>
```
```suggestion
// CHECK: %[[MASK_COMPRESSED:.+]] = arith.constant dense<[true, true, true, false]> : vector<4xi1>
```
?

banach-space commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir:
```diff
@@ -624,3 +624,28 @@ func.func @vector_maskedstore_i4_constant_mask(
+// CHECK: %[[ALLOC:.+]] = memref.alloc() : memref<12xi8>
+// CHECK: %[[MASK:.+]] = arith.constant dense<[false, true, true, true, true, false, false, false]> : vector<8xi1>
+// CHECK: %[[MASK2:.+]] = arith.constant dense<[true, true, true, false]> : vector<4xi1>
+// CHECK: %[[BITCAST:.+]] = vector.bitcast %[[PASSTHRU]] : vector<8xi4> to vector<4xi8>
```
```suggestion
// CHECK: %[[PTHRU_UPCAST:.+]] = vector.bitcast %[[PASSTHRU]] : vector<8xi4> to vector<4xi8>
```
?

lialan replied:
Nice! I am bad at naming as you can see.

banach-space commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir:
```diff
@@ -624,3 +624,28 @@ func.func @vector_maskedstore_i4_constant_mask(
+// CHECK: %[[C0:.+]] = arith.constant 0 : index
+// CHECK: %[[LOAD:.+]] = vector.maskedload %[[ALLOC]][%[[C0]]], %[[MASK2]], %[[BITCAST]]
+// CHECK-SAME : memref<12xi8>, vector<4xi1>, vector<4xi8> into vector<4xi8>
+// CHECK: %[[BITCAST2:.+]] = vector.bitcast %[[LOAD]] : vector<4xi8> to vector<8xi4>
```
```suggestion
// CHECK: %[[LOAD_DOWNCAST:.+]] = vector.bitcast %[[LOAD]] : vector<4xi8> to vector<8xi4>
```

banach-space commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:
```diff
@@ -249,3 +249,41 @@ func.func @vector_maskedload_i2_dynamic_indexing_mixed(%passthru: vector<3xi2>,
+
+// -----
+
+func.func @vector_maskedload_i4_constant_mask_unaligned(%passthru: vector<5xi2>) -> vector<5xi2> {
```
How is this actually different to `@vector_maskedload_i4_arith_constant` from the other file?

lialan replied:
Oh it is different in that the `numFrontPadElems` in the calculation is non zero in this case, so we check the affine map generation is correct.

banach-space reviewed (commented):
Thanks for the fix!

I think that the logic in `getCompressedMaskOp` could benefit from a high level refactor. In particular, why not replace:
```cpp
f (createMaskOp) {
} else if (constantMaskOp) {
} else if (constantOp) {
}
```

with (pseudo syntax):
```cpp
TypeSwitch<Operation *op>(maskOp).
Case<CreateMaskOp>[&](){}
Case<ConstantMaskOp>[&](){}
Case<ConstantOp>[&](){}
```

and then every case should be implemented in a dedicated hook.

More comments inline.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +77,137 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
 
-  assert(numFrontPadElems < numSrcElemsPerDest && "intraDataOffset must be less than scale");
+  assert(numFrontPadElems < numSrcElemsPerDest &&
+         "intraDataOffset must be less than scale");
```
%s/intraDataOffset/numFrontPadElem/

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +77,137 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+                if (numFrontPadElems == 0) {
+                  return rewriter.create<vector::ConstantMaskOp>(
+                      loc, newMaskType, newMaskDimSizes);
+                } else {
```
nit: no else after return

dcaballe reviewed (commented):
Doing a first pass for now. Thanks!

lialan commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -129,6 +132,45 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
       newMask = rewriter.create<arith::ConstantOp>(loc, newMaskType, denseAttr);
     }
+  } else if (constantOp) {
+    assert(shape.size() == 1 && "expected 1-D mask");
```
updated `shape` to `maskShape` to better represent the meaning.

Currently limiting it to only 1 d because the `arith.constant` op needs different treatment from the other two ops.

lialan replied:
I honestly don't know if there is any > 1-d masks being generated, but `arith.constant` does allow arbitrary dimension constants being generated in this way. Is there any mechanism guarding it off from any such cases before this pass?

cc @hanhanW @dcaballe

hanhanW replied:
I don't really know the answer.. My intuition is no in IREE project. My understanding is that all the n-D vector.transfer_read/write ops are unrolled to 1-D vector loads/stores on CPU path before we reach out to the emulation. GPU side, we don't use masking in IREE now.

However, I remember that @dcaballe added some n-D emulation support few months ago, so I'm not sure now.

I'd suggest to move this check before "TypeSwitch" and see what's happening. If it does not break anything, we don't need to make `arith.constant` case special. Otherwise, we leave it to where it is and add a TODO.

What do you think?

banach-space replied:
> I'd suggest to move this check before "TypeSwitch" and see what's happening. If it does not break anything, we don't need to make arith.constant case special. 

+1

No need to handle complexities that are not required. And if they do become required then at least we'll have a good motivating example to inform the design :)

lialan replied:
Turns out it breaks things as some masks indeed created by other kinds of ops do have multiple dimensions. I am leaving it as is with a TODO for now. I will come back later.

banach-space commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:
```diff
@@ -249,3 +249,41 @@ func.func @vector_maskedload_i2_dynamic_indexing_mixed(%passthru: vector<3xi2>,
+
+// CHECK: %[[CST0:.+]] = arith.constant dense<true> : vector<2xi1>
+// CHECK: %[[CST1:.+]] = arith.constant dense<0> : vector<8xi2>
+// CHECK: %[[INSERT:.+]] = vector.insert_strided_slice %[[PTH]], %[[CST1]]
+// CHECK-SAME: {offsets = [1], strides = [1]} : vector<5xi2> into vector<8xi2>
```
Isn't `INSERT` basically a "padded" version of `PTH`? Where the pad value is 0? Would you mind using more descriptive names? For example:

* `INSERT` -> `PTH_PADDED`
* `BCAST` -> `PTH_PADDED_UPCAST`
* `MASKLOAD` -> `LOAD` (IMO, this can be short)
* `BCAST2` -> `LOAD_DOWNCAST`
* `INSERT2` -> `MASK_PADDED`
* `EXTRACT` -> `RESULT`

These are just suggestions, feel free to re-use or propose sth else. Either way, it would be nice to make this self-documenting, similarly to `vector_maskedload_i4_arith_constant`.

banach-space commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:
```diff
@@ -249,3 +249,41 @@ func.func @vector_maskedload_i2_dynamic_indexing_mixed(%passthru: vector<3xi2>,
+  %mask = arith.constant dense<[false, true, true, true, false]> : vector<5xi1>
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %1 = vector.maskedload %0[%c1, %c0], %mask, %passthru :
```
IIUC, the key element of this example is the offset in the `vector.maskedload` operation? If that's the case, then it would be good to highlight the corresponding offset in the expected output. Either with a comment or clear LIT variable name. Also, using a higher value (e.g. "3", as a prime number, is nice) might be easier to illustrate what's happening here. "1" and "0" are super common and therefore it's hard to see them being propagated in tests.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +77,134 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+            // coverage size is mulitiple of bytes). The new mask will be like
+            // this before compressing:
+            //
+            // %new_mask = [false, false, true, false, true, false, false,
+            // false]
```
[nit]
```suggestion
            // %new_mask = 
            //   [false, false, true, false, true, false, false, false]
```

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +77,134 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+            // false]
+            auto denseAttr =
+                dyn_cast<DenseIntElementsAttr>(constantOp.getValue());
+            if (!denseAttr)
+              return std::nullopt;
```
Hm, is this case possible?

lialan replied:
ahh, I now use a `cast` to avoid such issue.

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +77,134 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+                              denseAttr.template value_end<bool>());
+            maskValues.resize(numElements * numSrcElemsPerDest, false);
+
+            // Compressing by combining every `numSrcElemsPerDest` elements:
```
```suggestion
            // Compressing by combining every block of `numSrcElemsPerDest` elements togehter:
```
?

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +77,134 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+            SmallVector<bool> maskValues(numFrontPadElems, false);
+            maskValues.append(denseAttr.template value_begin<bool>(),
+                              denseAttr.template value_end<bool>());
+            maskValues.resize(numElements * numSrcElemsPerDest, false);
```
`numElements` -> `numDestElements`

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +77,134 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+            if (!denseAttr)
+              return std::nullopt;
+            SmallVector<bool> maskValues(numFrontPadElems, false);
+            maskValues.append(denseAttr.template value_begin<bool>(),
+                              denseAttr.template value_end<bool>());
```
What's this meant to do?

lialan replied:
it is inserting the original mask values to the vector to form a zero-padded in the front and back mask value vector, which is used to form the new compressed make vector.

banach-space reviewed (commented):
Thanks for the updates!

I only see tests for `maskedload` - how about `maskedstore`?

lialan commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:
```diff
@@ -249,3 +249,41 @@ func.func @vector_maskedload_i2_dynamic_indexing_mixed(%passthru: vector<3xi2>,
+  return %1 : vector<5xi2>
+}
+
+// CHECK: func @vector_maskedload_i4_constant_mask_unaligned(
```
Well, let me do that in another PR then!

lialan commented:
@banach-space Thanks for the review! I've updated according to your suggestions. I also added a masked store test case to illustrate the effect. Given that `maskedstore` does not support unaligned emulation yet (which is on my todo list), I only added 1 of it (the aligned case). 

I think what really helps is to use more meaningful names in the test, it is quite self-explanatory!

hanhanW reviewed (approved):
LGTM, just a couple comments about comments. I'll help land the PR once we get an approval from @banach-space

banach-space commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type.mlir:
```diff
@@ -624,3 +648,29 @@ func.func @vector_maskedstore_i4_constant_mask(
+// CHECK-SAME: %[[VAL_TO_STORE:[a-zA-Z0-9]+]]:
+// CHECK: %[[ALLOC:.+]] = memref.alloc() : memref<20xi8>
+// CHECK: %[[MASK:.+]] = arith.constant dense<[false, true, true, true, true, true, false, false]> : vector<8xi1>
+// CHECK: %[[C12:.+]] = arith.constant 12 : index
```
[nit] 
```suggestion
// %c3 * 4 bits = 12
// CHECK: %[[IDX_FLATTENED:.+]] = arith.constant 12 : index
```

banach-space commented on mlir/lib/Dialect/Vector/Transforms/VectorEmulateNarrowType.cpp:
```diff
@@ -75,83 +75,133 @@ static FailureOr<Operation *> getCompressedMaskOp(OpBuilder &rewriter,
+            // this before compressing:
+            //
+            // %new_mask = [0, 0, 1, 0, 1, 0, 0, 0]
+            auto denseAttr = cast<DenseIntElementsAttr>(constantOp.getValue());
+            SmallVector<bool> paddedMaskValues(numFrontPadElems, false);
+            paddedMaskValues.append(denseAttr.template value_begin<bool>(),
+                                    denseAttr.template value_end<bool>());
```
Oh, now I see why I got confused 😅 `denseAttr` - > `originalMask` 🙏🏻 (I thought that `denseAttr` was referring to some abstract type)

banach-space commented:
> I think what really helps is to use more meaningful names in the test, it is quite self-explanatory!

Thank you for this comment and for updating the tests 🙏🏻. This makes reviewing and maintenance so much easier - our future selves will be very grateful! 😊

I’ve left a few small suggestions, but otherwise, this looks good to me (LGTM).

--- 
Given the growing complexity of this logic, I have a couple of high-level observations:

1. Should _"vector-emulate-narrow-type.mlir"_ be renamed to _"vector-emulate-narrow-type-aligned.mlir"_? Should the cases in _"vector-emulate-narrow-type-unaligned.mlir"_ essentially be variations of the cases in _"vector-emulate-narrow-type-aligned.mlir"_?

2. All the existing cases in "vector-emulate-narrow-type-unaligned.mlir" test `i2`, while the new cases test `i4`. In a PR like this, where the key "novelty" is support for `arith.constant` (i.e. a new Op), I’d normally expect the data types to remain consistent with the other tests. This way, it’s easier to isolate and understand the differences introduced by the new changes.

I’m not suggesting any additional large changes or refactors in this PR - @lialan has already done a great job! But I think the patterns/trends highlighted above emerge naturally and could serve as a helpful guiding principle for future PRs. If others agree, of course. 😊

banach-space reviewed (approved):
LGTM % some final nits

lialan commented on mlir/test/Dialect/Vector/vector-emulate-narrow-type-unaligned.mlir:
```diff
@@ -249,3 +249,41 @@ func.func @vector_maskedload_i2_dynamic_indexing_mixed(%passthru: vector<3xi2>,
+
+// -----
+
+func.func @vector_maskedload_i4_constant_mask_unaligned(%passthru: vector<5xi2>) -> vector<5xi2> {
```
@hanhanW copy/paste mis match, I should have double checked it. I will update it later in a batch refactoring.
